src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TransPatterns.java

```diff
@@ -301,8 +301,13 @@ public void visitDeconstructionPattern(JCDeconstructionPattern tree) {
                     extraTest = makeTypeTest(make.Ident(nestedTemp),
                                              make.Type(nested.type));
                 }
-            } else if (nested.type.isReference()) {
-                extraTest = makeBinary(Tag.NE, make.Ident(nestedTemp), makeNull());
+            }
+            else if (nested.type.isNullOrReference() || nested.type.isPrimitive()) {
+                if(nested instanceof JCDeconstructionPattern) {
+                    extraTest = makeBinary(Tag.NE, make.Ident(nestedTemp), makeNull());
+                }
+                else
+                    extraTest = makeLit(syms.booleanType, 1);
```

Comment:
Thanks for working on this.

A few general comments here:
- we usually put `else` at the same line as the preceding closing `}`, e.g. `} else`.
- we usually put `{}` around the else section when there are brackets around the then section
- we normally don't (currently) use `instanceof` to check AST node types - we usually use `JCTree.hasTag` or `JCTree.getTag`.

Here, seems this would work as well, right?
```java
            } else if (nested.type.isReference() && nested.hasTag(Tag.DECONSTRUCTIONPATTERN)) {
                extraTest = makeBinary(Tag.NE, make.Ident(nestedTemp), makeNull());
            }
```

Reply:
Thank you @lahodaj 

Addressed.